### PR TITLE
[MM-44363] Detect M1 and load compatible Elasticsearch version

### PR DESCRIPTION
**Pull request by noxer**
#### Summary
The PR adds functionality to the Makefile to detect M1 Macs. When an M1 is detected, the Makefile tries to load a compatible Elasticsearch version (if required).

#### Ticket Link
[MM-44363](https://mattermost.atlassian.net/browse/MM-44363)

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by agnivade** [comment]
Were we able to find out why it gets stuck in non-M1 laptops?

**Comment by noxer** [comment]
> Were we able to find out why it gets stuck in non-M1 laptops?

There are a multitude of reasons why Elasticsearch would not properly start. I'd like to put that investigation into it's own ticket. I feel like fixing this for M1 users is more pressing now.

**Comment by isacikgoz** [comment]
It seems like there are images built for Apple silicon macs for elasticsearch. Since we can detect the M1/M2 macs, could we just use a different base image for elasticsearch? We would need to create another image with an `arm` tag and use it instead. 

We would need to create the tag here I guess: https://github.com/mattermost/mattermost-elasticsearch-docker

The following image works on a M1 mac for instance:

```bash
docker run -p 9200:9200 -p 9300:9300 -e "discovery.type=single-node" docker.elastic.co/elasticsearch/elasticsearch:7.17.3-arm64
```

**Comment by noxer** [comment]
Yes that could work. We can replace the elasticsearch container with a docker-compose overlay. We just need to create a branch in the [mattermost-elasticsearch-docker](https://github.com/mattermost/mattermost-elasticsearch-docker) repo and tag it accordingly.

**Comment by agnivade** [comment]
Agree. Tim, let me know if you need any help with the release team. Feel free to ask around in the Team: Release channel in the private team.

**Review by agnivade — APPROVED** [review]

File: docker-compose.makefile.m1.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+version: '2.4'
+services:
+  elasticsearch:
+  image: 'mattermost/mattermost-elasticsearch-docker:7.17.3-arm64v8'
```
Comment (isacikgoz) [inline docker-compose.makefile.m1.yml:4]: Out of curiosity; Is it a hard requirement to specify `arm64` version? The examples I saw was using `arm64` generally..

**Review by isacikgoz — APPROVED** [review]
LGTM, just a question below 👇

File: docker-compose.makefile.m1.yml (hunk 1 above)
Comment (noxer) [reply docker-compose.makefile.m1.yml:4]: I took it from here: https://hub.docker.com/_/elasticsearch (under "Supported architectures")
When I started, I wanted to use 'arm64' as well but then thought about possible other ARM architectures we might support in the future and decided to be more specific. It's probably unnecessary but it's no effort to add the `v8` now but may be more complicated in the future.

**Review by phoinixgrr — CHANGES_REQUESTED** [review]
This fails for me with 

```
❯ make start-docker                                         
M1 detected, applying elasticsearch override
Starting docker containers
docker-compose rm start_dependencies
No stopped containers
go run ./build/docker-compose-generator/main.go mysql postgres inbucket  minio openldap elasticsearch | docker-compose -f docker-compose.makefile.yml -f /dev/stdin -f docker-compose.makefile.m1.yml  run -T --rm start_dependencies
yaml: line 4: did not find expected key
make: *** [start-docker] Error 15
❯ uname -a
Darwin pedobear 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:37 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T6000 arm64
```

**Comment by phoinixgrr** [comment]
Moreover

The [proposed way](https://github.com/docker/docker.github.io/blob/master/compose/compose-file/compose-file-v2.md#platform) to approach this, is to specify the `platform` directive, overiding `services.elasticsearch.platform` 

```
  elasticsearch:
    platform: linux/arm64/v8 
```


Also, the specific platform build, should be available in the repo used
Even though in the upstream repo is available: 
   
Screenshot 2022-06-15 at 12 31 09

In the repo used here is not 
https://hub.docker.com/r/mattermost/mattermost-elasticsearch-docker/tags

Screenshot 2022-06-15 at 12 32 28



--- 

Finally, In my testing, I was able to able to make this run with the following diff as a POC: 

```diff
diff --git a/build/docker-compose.common.yml b/build/docker-compose.common.yml
index 6d15b22fe..f7f30016f 100644
--- a/build/docker-compose.common.yml
+++ b/build/docker-compose.common.yml
@@ -84,7 +84,7 @@ services:
       LDAP_DOMAIN: "mm.test.com"
       LDAP_ADMIN_PASSWORD: "mostest"
   elasticsearch:
-    image: "mattermost/mattermost-elasticsearch-docker:7.17.3"
+    image: "elasticsearch:7.17.3"
     networks:
       - mm-test
     environment:
diff --git a/docker-compose.makefile.yml b/docker-compose.makefile.yml
index eb10dc138..1cca166c4 100644
--- a/docker-compose.makefile.yml
+++ b/docker-compose.makefile.yml
@@ -53,6 +53,7 @@ services:
         service: openldap
   elasticsearch:
     restart: 'no'
+    platform: linux/arm64/v8
     container_name: mattermost-elasticsearch
     ports:
       - "9200:9200"
(END)
```

**Comment by agnivade** [comment]
Thanks for the help Akis!

**Review by agnivade — APPROVED** [review]

**Comment by noxer** [comment]
The state is not optimal yet @agnivade . It skips installing an ES plugin (because we are not using the mm-es-docker image.)

**Comment by agnivade** [comment]
Understood. Is the plugin necessary for MM to run? Feel free to take a call. You have my approval already :)

**Comment by phoinixgrr** [comment]
Still fails for me with: 

```
 ❯ make start-docker
M1 detected, applying elasticsearch override
Starting docker containers
docker-compose rm start_dependencies
No stopped containers
go run ./build/docker-compose-generator/main.go mysql postgres inbucket  minio openldap elasticsearch | docker-compose -f docker-compose.makefile.yml -f /dev/stdin -f docker-compose.makefile.m1.yml  run -T --rm start_dependencies
yaml: line 4: did not find expected key
make: *** [start-docker] Error 15
```


Proposed fix: 
```diff
diff --git a/docker-compose.makefile.m1.yml b/docker-compose.makefile.m1.yml
index 88272fa14..522e418bd 100644
--- a/docker-compose.makefile.m1.yml
+++ b/docker-compose.makefile.m1.yml
@@ -1,7 +1,7 @@
 version: '2.4'
 services:
   elasticsearch:
-  image: "elasticsearch:7.17.3"
+    image: "elasticsearch2:7.17.3"
     platform: linux/arm64/v8
     restart: 'no'
     container_name: mattermost-elasticsearch
```

**Comment by noxer** [comment]
@phoinixgrr Thank you. Sorry I missed that. :facepalm:

**Review by phoinixgrr — APPROVED** [review]
@noxer  No worries. Great work! 🎖️ 
 :shipit:

**Pull request merged**